### PR TITLE
schema: Replace 'anyOf' with 'oneOf'.

### PR DESCRIPTION
**Pull request by IvanGoncharov**
Simplified schema to show problem:

``` json
{
  "anyOf": [
    {
      "type": "boolean"
    },
    {
      "type": "object"
    }
  ]
}
```

It obvious that this two subschema couldn't be true at the same time.
So `anyOf` should be replaced with `oneOf`.

Such inconsistency confuses readers of `schema`, and serves as a bad example.
Moreover, it produces confusing errors when you validate your schemas.

### CODE REVIEW

**Comment by JanTvrdik** [comment]
Note that `anyOf` validation is usually faster than `oneOf`.

**Comment by Ixrec** [comment]
The argument given here seems to be assuming that there are no valid use cases for anyOf where oneOf will not do. Although I've not personally run into such a use case in practice, it's easy to think of one:
```
{
	"anyOf": [
		{
			"type": "object",
			"properties": {
				"a": { "type": "string" },
				"b": { "type": "string" }
			},
			"required": ["a", "b"]
		},
		{
			"type": "object",
			"properties": {
				"x": { "type": "string" },
				"y": { "type": "string" }
			},
			"required": ["x", "y"]
		}
	]
}

...

{ x: "" } // FAIL
{ y: "" } // FAIL
{ x: "", y: "" } // PASS
{ a: "", x: "" } // FAIL
{ a: "", x: "", y: "" } // PASS
{ a: "", b: "", y: "" } // PASS
{ a: "", q: "", y: "" } // FAIL
{ a: "", b: "", x: "", y: "" } // PASS
{ a: "", b: "", x: "", y: "", q: "" } // PASS
```
While it's easy to convert this short example to oneOf by adding the a/b/x/y combination as a third subschema, if I added more properties or more subschemas the oneOf equivalent would suffer a combinatorial explosion while anyOf would not.

I have no idea whether schemas like this are remotely common in practice, but since they at least exist in principle, and they're already in the standard, the strongest "anti-anyOf" stance that seems reasonable to me without further user feedback/polling would be "recommend that schema writers always use oneOf unless they're sure they need anyOf".

**Comment by IvanGoncharov** [comment]
@JanTvrdik I understand, but it's a hack. And I don't think improvement is so big since it's only two schemas and validation fail will occur on `type` keyword. If validator optimized to check `type` first(it's good generic optimization) than overhead is only this check.

**Comment by jdesrosiers** [comment]
@IvanGoncharov I find `anyOf` not only more efficient, but also more intuitive that `oneOf` in this case (and most cases).  Here is why.

JSON Schema is about adding constraints.  The empty schema validates against the full set of JSON documents.  Every time you add a keyword to the schema, you trim the set of valid JSON documents.  When using `anyOf`, it adds the constraint that one of the schemas must be valid.  `oneOf` works the same as `anyOf` with the additional constraint that the match must be mutually exclusive.  Since the schemas are naturally mutually exclusive, adding a constraint to enforce that constraint is illogical.  It's over specificataion.  Take the following example.

```json
{
  "type": "string",
  "enum": ["foo", "bar"]
}
```

This schema is over specified in the same way as your example except that it takes two keywords to do it rather than one.  In this case the `type` keyword is redundant because the value must be a string in order to validate against either "foo" or "bar".

I do agree that the names chosen for `anyOf` and `oneOf` are a bit too ambiguous especially for those new to  JSON Schema.  This is especially true because JSON Schema uses a paradigm that is unfamiliar to most people.  Everyone is always looking for OO parallels that just aren't there.  But that isn't relevant to this question, so I digress.

In conclusion, the schema is using the right keyword.  I wouldn't advocate changing it.  However, I could get behind a renaming of the keywords in future draft if someone comes up with something that better communicates their semantics.  But, I suspect education is the best path forward with this issue.
